widget/embedded/src/components/SwapsGroup/SwapsGroup.tsx

```diff
@@ -2,15 +2,60 @@ import type { PropTypes } from './SwapsGroup.types';
 
 import { i18n } from '@lingui/core';
 import { Divider, SwapListItem, Typography } from '@rango-dev/ui';
-import React from 'react';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { GroupedVirtuoso } from 'react-virtuoso';
 
 import { limitDecimalPlaces } from '../../utils/numbers';
 
-import { Group, groupStyles, SwapList, Time } from './SwapsGroup.styles';
+import {
+  Group,
+  groupStyles,
+  SwapItemContainer,
+  SwapList,
+  Time,
+} from './SwapsGroup.styles';
+
+const ITEMS_PER_PAGE = 10;
 
 export function SwapsGroup(props: PropTypes) {
   const { list, onSwapClick, groupBy, isLoading } = props;
-  const groups = groupBy ? groupBy(list) : [{ title: 'History', swaps: list }];
+  const [currentGroupCounts, setCurrentGroupCounts] = useState<number[]>([]);
+  const loadedItems = useRef(0);
+  const { swaps, groupCounts, groups } = groupBy(list);
+
+  const calculateGroupsSoFar = useCallback(
+    (totalGroups: number[], count: number) => {
+      const groups: number[] = [];
+      let index = 0;
+      do {
+        const group = totalGroups[index];
+        groups.push(Math.min(group, count));
+        count -= group;
+        index++;
+      } while (count > 0 && index <= totalGroups.length);
+      return groups;
+    },
+    []
+  );
+
+  const loadMore = useCallback(() => {
+    if (list.length - loadedItems.current > ITEMS_PER_PAGE) {
```

Comment:
Can we have a better approach such as Math.min or something else instead of these conditions?